nexus/src/app/background/tasks/decommissioned_disk_cleaner.rs

```diff
@@ -0,0 +1,458 @@
+//!
+//! This cannot happen at decommissioning time, because it depends
+//! on region replacement, which happens in the backgorund.
+//!
+//! Cleanup involves deleting database records for disks (datasets, zpools)
+//! that are no longer viable after the physical disk has been decommissioned.
+
+use crate::app::background::BackgroundTask;
+use anyhow::Context;
+use futures::future::BoxFuture;
+use futures::FutureExt;
+use nexus_db_model::Zpool;
+use nexus_db_queries::context::OpContext;
+use nexus_db_queries::db::pagination::Paginator;
+use nexus_db_queries::db::DataStore;
+use nexus_types::identity::Asset;
+use omicron_common::api::external::Error;
+use omicron_uuid_kinds::{GenericUuid, ZpoolUuid};
+use std::num::NonZeroU32;
+use std::sync::Arc;
+
+/// Background task that cleans decommissioned disk DB records.
+pub struct DecommissionedDiskCleaner {
+    datastore: Arc<DataStore>,
+    disable: bool,
+}
+
+#[derive(Debug, Default)]
+struct ActivationResults {
+    found: usize,
+    not_ready_to_be_deleted: usize,
+    deleted: usize,
+    error_count: usize,
+}
+
+const MAX_BATCH: NonZeroU32 = unsafe {
+    // Safety: last time I checked, 100 was greater than zero.
+    NonZeroU32::new_unchecked(100)
+};
+
+impl DecommissionedDiskCleaner {
+    pub fn new(datastore: Arc<DataStore>, disable: bool) -> Self {
+        Self { datastore, disable }
+    }
+
+    async fn clean_all(
+        &mut self,
+        results: &mut ActivationResults,
+        opctx: &OpContext,
+    ) -> Result<(), anyhow::Error> {
+        slog::info!(opctx.log, "Decommissioned disk cleaner running");
+
+        let mut paginator = Paginator::new(MAX_BATCH);
+        let mut last_err = Ok(());
+        while let Some(p) = paginator.next() {
+            let zpools = self
+                .datastore
+                .zpool_on_decommissioned_disk_list(
+                    opctx,
+                    &p.current_pagparams(),
+                )
+                .await
+                .context("failed to list zpools on decommissioned disks")?;
+            paginator = p.found_batch(&zpools, &|zpool| zpool.id());
+            self.clean_batch(results, &mut last_err, opctx, &zpools).await;
+        }
+
+        last_err
+    }
+
+    async fn clean_batch(
+        &mut self,
+        results: &mut ActivationResults,
+        last_err: &mut Result<(), anyhow::Error>,
+        opctx: &OpContext,
+        zpools: &[Zpool],
+    ) {
+        results.found += zpools.len();
+        slog::debug!(opctx.log, "Found zpools on decommissioned disks"; "count" => zpools.len());
+
+        for zpool in zpools {
+            let zpool_id = ZpoolUuid::from_untyped_uuid(zpool.id());
+            slog::trace!(opctx.log, "Deleting Zpool"; "zpool" => %zpool_id);
+
+            match self
+                .datastore
+                .zpool_delete_self_and_all_datasets(opctx, zpool_id)
+                .await
+            {
+                Ok(_) => {
+                    slog::trace!(
```

Comment:
if I was looking at logs, I'd prefer to see these messages over the "not ready for deletion" below. The "not ready for deletion" message will repeat until region + snapshot replacement has occurred, so that could be `trace` instead.

Reply:
Totally agreed, I'll swap these log levels!